File: backend/api/chatbot.py
from flask import Flask, request, jsonify
from flask_cors import CORS
import mysql.connector
import json
import os
from datetime import datetime
import openai
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

# Flask app setup
app = Flask(__name__)
CORS(app)

# OpenAI API key from environment
openai.api_key = os.getenv("OPENAI_API_KEY")

# Database configuration
config = {
    'user': os.getenv('DB_USER'),
    'password': os.getenv('DB_PASSWORD'),
    'host': os.getenv('DB_HOST'),
    'database': os.getenv('DB_NAME')
}

def get_db_connection():
    return mysql.connector.connect(**config)

def analyze_symptoms(symptoms):
    try:
        response = openai.ChatCompletion.create(
            model="gpt-3.5-turbo",
            messages=[
                {"role": "system", "content": "You are a medical assistant. Analyze the following symptoms and provide a structured response. Include possible conditions, severity, advice, medications, and precautions."},
                {"role": "user", "content": symptoms}
            ],
            temperature=0.7,
            max_tokens=500
        )
        analysis = response.choices[0].message.content

        return {
            'type': 'analysis',
            'message': 'Based on your symptoms, here\'s what I found:',
            'analysis': analysis,
            'timestamp': datetime.now().isoformat()
        }

    except Exception as e:
        print(f"Error in symptom analysis: {str(e)}")
        return {
            'type': 'error',
            'message': 'Sorry, I encountered an error while analyzing your symptoms.',
            'timestamp': datetime.now().isoformat()
        }

@app.route('/api/chatbot', methods=['POST'])
def chatbot():
    try:
        data = request.get_json()
        if not data or 'message' not in data:
            return jsonify({'error': 'No message provided'}), 400

        user_message = data['message']
        response = analyze_symptoms(user_message)

        # Log to database
        try:
            conn = get_db_connection()
            cursor = conn.cursor()
            cursor.execute("""
                INSERT INTO chat_logs (user_message, bot_response, created_at)
                VALUES (%s, %s, %s)
            """, (user_message, json.dumps(response), datetime.now()))
            conn.commit()
            cursor.close()
            conn.close()
        except Exception as e:
            print(f"Database error: {str(e)}")

        return jsonify(response)

    except Exception as e:
        print(f"Error in chatbot endpoint: {str(e)}")
        return jsonify({'error': 'Internal server error'}), 500

if __name__ == '__main__':
    app.run(debug=os.getenv('FLASK_DEBUG', '1') == '1', port=5000)

File: backend/api/chatbot.php
<?php
header('Content-Type: application/json');
require_once '../db.php';

// Function to analyze symptoms and provide response
function analyzeSymptoms($symptoms) {
    // Convert symptoms to lowercase for better matching
    $symptoms = strtolower($symptoms);
    
    // Common symptoms and their possible conditions
    $symptomPatterns = [
        'fever' => [
            'condition' => 'Fever',
            'severity' => 'moderate',
            'advice' => 'Monitor your temperature. If it exceeds 38°C (100.4°F) or persists for more than 3 days, consult a doctor.',
            'medications' => ['Paracetamol', 'Ibuprofen'],
            'precautions' => ['Stay hydrated', 'Get plenty of rest', 'Use a cold compress']
        ],
        'headache' => [
            'condition' => 'Headache',
            'severity' => 'mild',
            'advice' => 'Rest in a quiet, dark room. If pain persists or worsens, seek medical attention.',
            'medications' => ['Paracetamol', 'Ibuprofen'],
            'precautions' => ['Stay hydrated', 'Avoid bright lights', 'Reduce screen time']
        ],
        'cough' => [
            'condition' => 'Cough',
            'severity' => 'mild',
            'advice' => 'If cough persists for more than 2 weeks or is accompanied by fever, consult a doctor.',
            'medications' => ['Cough syrup', 'Honey and lemon'],
            'precautions' => ['Stay hydrated', 'Use a humidifier', 'Avoid irritants']
        ],
        'stomach pain' => [
            'condition' => 'Stomach Pain',
            'severity' => 'moderate',
            'advice' => 'If pain is severe or accompanied by vomiting/diarrhea, seek medical attention.',
            'medications' => ['Antacids', 'Pepto-Bismol'],
            'precautions' => ['Eat light foods', 'Stay hydrated', 'Avoid spicy foods']
        ]
    ];

    // Check for emergency keywords
    $emergencyKeywords = ['chest pain', 'difficulty breathing', 'severe pain', 'unconscious', 'bleeding'];
    foreach ($emergencyKeywords as $keyword) {
        if (strpos($symptoms, $keyword) !== false) {
            return [
                'type' => 'emergency',
                'message' => 'This appears to be an emergency. Please call emergency services immediately or go to the nearest hospital.',
                'condition' => 'Emergency',
                'severity' => 'critical',
                'advice' => 'Seek immediate medical attention.',
                'medications' => [],
                'precautions' => ['Call emergency services', 'Stay calm', 'Follow emergency operator instructions']
            ];
        }
    }

    // Match symptoms with known patterns
    $matchedConditions = [];
    foreach ($symptomPatterns as $pattern => $data) {
        if (strpos($symptoms, $pattern) !== false) {
            $matchedConditions[] = $data;
        }
    }

    if (empty($matchedConditions)) {
        return [
            'type' => 'general',
            'message' => 'I understand you\'re experiencing symptoms. Could you please provide more details about your symptoms?',
            'condition' => 'Unknown',
            'severity' => 'unknown',
            'advice' => 'Please provide more specific symptoms for better analysis.',
            'medications' => [],
            'precautions' => []
        ];
    }

    // If multiple conditions are matched, combine the information
    $response = [
        'type' => 'analysis',
        'message' => 'Based on your symptoms, here\'s what I found:',
        'conditions' => $matchedConditions,
        'general_advice' => 'Monitor your symptoms and seek medical attention if they worsen.',
        'precautions' => ['Stay hydrated', 'Get plenty of rest', 'Monitor symptoms']
    ];

    return $response;
}

// Handle POST request
if ($_SERVER['REQUEST_METHOD'] === 'POST') {
    $data = json_decode(file_get_contents('php://input'), true);
    
    if (!isset($data['message'])) {
        echo json_encode(['error' => 'No message provided']);
        exit;
    }

    $userMessage = $data['message'];
    $response = analyzeSymptoms($userMessage);

    // Log the interaction in the database
    try {
        $stmt = $pdo->prepare("INSERT INTO chat_logs (user_message, bot_response, created_at) VALUES (?, ?, NOW())");
        $stmt->execute([$userMessage, json_encode($response)]);
    } catch (PDOException $e) {
        // Log error but don't fail the request
        error_log("Database error: " . $e->getMessage());
    }

    echo json_encode($response);
} else {
    echo json_encode(['error' => 'Invalid request method']);
}
?> 